src/ql/pass/opt/const_prop/detail/propagate.cc

```diff
@@ -0,0 +1,224 @@
+        handle_expression(if_else_branch.condition);
+
+        // descend branch body
+        if_else_branch.body->visit(*this);
+    }
+
+    /**
+     * Visitor function for `FunctionCall` nodes.
+     */
+    void visit_function_call(ir::FunctionCall &function_call) override {
+        DEBUG("recursing into operands of function call '" << ir::describe(function_call) << "'");
+        for (auto &operand : function_call.operands) {
+            handle_expression(operand);
+        }
+    }
+
+    /**
+     * Visitor function for `FunctionCall` nodes.
+     */
+    void visit_expression(ir::Expression &expression) override {
+        if(auto function_call = expression.as_function_call()) {
+            DEBUG("recursing into function call '" << ir::describe(*function_call) << "'");
+            visit_function_call(*function_call);   // descend
+        }
+    }
+
+private:    // types
+
+    // function pointer
+    typedef FncRet (*tOpFunc)(const ir::Ref &ir, const FncArgs &args);
+
+    // map function key (see register_functions()) to function pointer
+    using FuncMap = std::map<utils::Str, tOpFunc>;
+
+private:    // vars
+    const ir::Ref &ir;
+    FuncMap func_map;
+
+private:    // functions
+
+    /*
+     * Handle an expression node, i.e. replace eligible functions calls with a literal expression.
+     *
+     * Note that we cannot directly use visit_expression(), because that has a 'ir::Expression &' as parameter.
+     * It is therefore not possible to change a function_call into a (say) int_literal. Here the expression is
+     * wrapped in a 'utils::One', allowing polymorphic access. This does force us however to visit all relevant node
+     * types that contain an expression (although we can skip e.g. custom_instruction.operands).
+     */
+    void handle_expression(utils::One<ir::Expression> &expression) {
+        DEBUG("descending '" << ir::describe(expression));
+        visit_expression(*expression);    // descend
+        DEBUG("done descending '" << ir::describe(expression));
+
+        if (auto function_call = expression->as_function_call()) {
+            QL_IOUT("function call '" << ir::describe(*function_call) << "'");
+
+            // generate key, consistent with register_functions()
+            utils::Str key = function_call->function_type->name + "_";
+            for (auto &operand : function_call->operands) {
+                if (operand->as_int_literal()) {
+                    key += "i";
+                } else if (operand->as_bit_literal()) {
+                    key += "b";
+                } else {
+                    DEBUG("not touching operand '" << ir::describe(operand) << "'");
+                    return;     // we don't handle functions with other operand types, but leave them untouched
+                }
+            }
+
+            // NB: we don't perform type promotions like libqasm's cQASM resolver, see FunctionTable::call
```

Comment:
Why not?

Reply:
The basic reason is that it currently makes little sense since OpenQL only supports a subset of the types libqasm knows about. We could promote a bool to an int when possible, but since (at least in the CC) int's and bool's live in different memory spaces requiring an explicit int cast by the user seems a better solution

Reply:
There is a cast from bit to int (`int(bit b)`) in the standard platform functions, which involves moving between the two memory spaces on the CC. No other casts exist by default, although `operator>` could be used to get a C-style int to bool conversion.
On the CC, bits live in Distributed Shared Memory, which can only be written by memory, not by instructions (or in the STATE register, which **can** be witten to, but which isn't supported yet by the backend)